src/test/java/com/connexta/ingest/service/impl/IngestServiceImplUnitTests.java

```diff
@@ -0,0 +1,322 @@
+  }
+
+  /* END ingest MetacardStorageAdaptor tests */
+
+  /* START ingest retrieveEndpoint tests */
+
+  @Test
+  public void testIngestWhenRetrieveEndpointCreatesInvalidURI() throws Exception {
+    // given
+    final IngestService ingestService =
+        new IngestServiceImpl(
+            mockStoreClient,
+            mockMetacardStorageAdaptor,
+            "this will cause the metacardLocation to be an invalid URI",
+            mockTransformClient);
+
+    final Long fileSize = 1L;
+    final String mimeType = "mimeType";
+    final InputStream inputStream = mock(InputStream.class);
+    final String fileName = "fileName";
+    final long metacardFileSize = 1L;
+    final InputStream metacardInputStream = mock(InputStream.class);
+    when(mockStoreClient.store(fileSize, mimeType, inputStream, fileName))
+        .thenReturn(new URI("http://localhost:9041/mis/product/1234"));
+
+    // when
+    final StoreMetacardException storeMetacardException =
+        assertThrows(
+            StoreMetacardException.class,
+            () ->
+                ingestService.ingest(
+                    fileSize,
+                    mimeType,
+                    inputStream,
+                    fileName,
+                    metacardFileSize,
+                    metacardInputStream));
+
+    // then
+    assertThat(storeMetacardException.getReason(), equalTo("Unable to construct retrieve URI"));
+  }
+
+  /* END ingest retrieveEndpoint tests */
+
+  /* START ingest TransformClient tests */
+
+  @Test
+  public void testIngestWhenTransformClientThrowsTransformException() throws Exception {
+    // given
+    final IngestService ingestService =
+        new IngestServiceImpl(
+            mockStoreClient,
+            mockMetacardStorageAdaptor,
+            "http://localhost:9040/ingest/",
+            mockTransformClient);
+
+    final Long fileSize = 1L;
+    final String mimeType = "mimeType";
+    final InputStream inputStream = mock(InputStream.class);
+    final String fileName = "fileName";
+    final long metacardFileSize = 1L;
+    final InputStream metacardInputStream = mock(InputStream.class);
+    final URI location = new URI("http://localhost:9041/mis/product/1234");
+    when(mockStoreClient.store(fileSize, mimeType, inputStream, fileName)).thenReturn(location);
+
+    final TransformException transformException = mock(TransformException.class);
+    doThrow(transformException)
+        .when(mockTransformClient)
+        .requestTransform(eq(location), eq(mimeType), any(URI.class));
+
+    // when
+    final TransformException thrown =
+        assertThrows(
+            TransformException.class,
+            () ->
+                ingestService.ingest(
+                    fileSize,
+                    mimeType,
+                    inputStream,
+                    fileName,
+                    metacardFileSize,
+                    metacardInputStream));
+
+    // then
+    assertThat(thrown, is(transformException));
```

Comment:
:question:  Is the `assertThat(thrown, is(transformException));` necessary? The `assertThrows` already checks if the class of the exception is correct.

Reply:
Yeah, I think this is just a duplicate

Reply:
I want to verify that the exception thrown is the exact exception that was thrown by the TransformClient. It's not good enough to just check that it's the expected type. We use the information in the TransformException to form the ErrorResponse, so it has to be correct. This test should break, for example, if we catch the TransformException thrown by the TransformClient and throw a brand new TransformException because the new exception would contain the right cause/details.